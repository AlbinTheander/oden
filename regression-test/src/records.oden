package records/main

import fmt

type Foo = { name: string }
type Bar = { foo: Foo }

// Construct nested record, type is inferred.
bar = {
  foo = {
    name = "mono"
  }
}

// Accepts nested record, as defined by 'Bar' and 'Foo' above.
printFooName :: Bar -> ()
printFooName(x) = fmt.Println(x.foo.name)

// Accept any record with a name field.
anyTypeOfName :: forall a. a -> { name: a }
anyTypeOfName(name) = { name = name }

// Construct a record with a string name.
poly = anyTypeOfName("poly")

// Return name field from any record type with at least that field. Type is
// inferred.
getName(x) = x.name

main() = {
  printFooName(bar)
  fmt.Println(poly.name)
  fmt.Println(getName({
    name = "passing a record like this",
    age = 42
  }))
}
