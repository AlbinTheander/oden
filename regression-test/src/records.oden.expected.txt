mono
poly
passing a record like this
